server/src/main/java/org/elasticsearch/action/search/LookupJoinHitSearchPhase.java

```diff
@@ -0,0 +1,252 @@
+                    if (matchValue == null) {
+                        continue;
+                    }
+                    final String matchId = matchValue.toString();
+                    if (fetchedJoinHits.stream().anyMatch(j -> j.getId().equals(matchId))) {
+                        continue;
+                    }
+                    final SearchHit.JoinHit cachedHit = cachedJoinHits.get(matchId);
+                    if (cachedHit != null) {
+                        leftHit.addJoinHit(group.query.getName(), cachedHit);
+                    } else {
+                        group.leftHits.computeIfAbsent(matchId, k -> new ArrayList<>()).add(leftHit);
+                    }
+                }
+            }
+        }
+        return groups.stream()
+            .flatMap(g -> g.leftHits.entrySet().stream().map(e -> new LookupRequest(g.query, e.getKey(), e.getValue())))
+            .collect(Collectors.toList());
+    }
+
+    static void mergeJoinHits(List<LookupRequest> lookupRequests, MultiGetItemResponse[] rightResponses) {
+        assert lookupRequests.size() == rightResponses.length : lookupRequests.size() + " != " + rightResponses.length;
+        for (int i = 0; i < lookupRequests.size(); i++) {
+            final MultiGetItemResponse rightResponse = rightResponses[i];
+            final LookupRequest lookupRequest = lookupRequests.get(i);
+            final SearchHit.JoinHit joinHit;
+            if (rightResponse.getFailure() != null) {
+                joinHit = new SearchHit.JoinHit(lookupRequest.matchId, rightResponse.getFailure().getFailure());
+            } else {
+                joinHit = joinHitFromGetResult(lookupRequest, rightResponse.getResponse().getResult());
+            }
+            for (SearchHit leftHit : lookupRequest.getLeftHits()) {
+                leftHit.addJoinHit(lookupRequest.query.getName(), joinHit);
+            }
+        }
+    }
+
+    static SearchHit.JoinHit joinHitFromGetResult(LookupRequest request, GetResult getResult) {
+        if (getResult.isExists()) {
+            assert Objects.equals(request.getMatchId(), getResult.getId()) : request.getMatchId() + " != " + getResult.getId();
+            return new SearchHit.JoinHit(getResult.getId(), getResult.getDocumentFields(), getResult.internalSourceRef());
+        } else {
+            return new SearchHit.JoinHit(request.getMatchId(),
+                new ResourceNotFoundException(
+                    "join hit for id [" + request.getMatchId() + "] on index [" + request.getIndex() + "] doesn't exist"));
+        }
+    }
+
+    /**
+     * An optimization of {@link org.elasticsearch.action.search.LookupJoinHitSearchPhase} that tries to lookup join hits from shards
+     * that are allocated on this node to avoid network requests if these lookup requests are executed on the coordinating node.
+     */
+    public static void tryLookupJoinHitsLocally(SearchContext searchContext, SearchHits searchHits) {
```

Comment:
shouldn't the same be possible by just specifying "preference=_local" on the get request?

Reply:
Great point. We can achieve the same thing with `preference=_only_local`. I am on it.